lib/fog/orchestration/util/recursive_hot_file_loader.rb

```diff
@@ -0,0 +1,227 @@
+          #  If it's an Hash, we deep_copy it so the passed argument
+          #  is not modified by get_file_contents.
+          template = deep_copy(template)
+          @visited = Set.new
+          @files = files || {}
+          @template = get_template_contents(template)
+        end
+
+        # Return string
+        def url_join(prefix, suffix)
+          if prefix
+            # URI.join replaces prefix parts before a
+            #  trailing slash. See https://docs.ruby-lang.org/en/2.3.0/URI.html.
+            prefix += '/' unless prefix.to_s.end_with?("/")
+            suffix = URI.join(prefix, suffix)
+            # Force URI to use traditional file scheme representation.
+            suffix.host = "" if suffix.scheme == "file"
+          end
+          suffix.to_s
+        end
+
+        # Retrieve a template content.
+        #
+        # @param template_file can be either:
+        #          - a raw_template string
+        #          - an URI string
+        #          - an Hash containing the parsed template.
+        #
+        # XXX: we could use named parameters
+        # and better mimic heatclient implementation.
+        def get_template_contents(template_file)
+          Fog::Logger.debug("get_template_contents [#{template_file}]")
+
+          raise "template_file should be Hash or String, not #{template_file.class.name}" unless
+            template_file.kind_of?(String) || template_file.kind_of?(Hash)
+
+          local_base_url = url_join("file:/", File.absolute_path(Dir.pwd))
+
+          if template_file.kind_of?(Hash)
+            template_base_url = local_base_url
+            template = template_file
+          elsif template_is_raw?(template_file)
+            template_base_url = local_base_url
+            template = YAML.safe_load(template_file, [Date])
+          elsif template_is_url?(template_file)
+            template_file = normalise_file_path_to_url(template_file)
+            template_base_url = base_url_for_url(template_file)
+            raw_template = read_uri(template_file)
+            template = YAML.safe_load(raw_template, [Date])
+
+            Fog::Logger.debug("Template visited: #{@visited}")
+            @visited.add(template_file)
+          else
+            raise "template_file is not a string of the expected form"
+          end
+
+          get_file_contents(template, template_base_url)
+
+          template
+        end
+
+        # Traverse the template tree looking for get_file and type
+        #   and populating the @files attribute with their content.
+        #   Resource referenced by get_file and type are eventually
+        #   replaced with their absolute URI as done in heatclient
+        #   and shade.
+        #
+        def get_file_contents(from_data, base_url = nil)
+          Fog::Logger.debug("Processing #{from_data} with base_url #{base_url}")
+
+          # Recursively traverse the tree.
+          if recurse_if(from_data)
```

Comment:
We don't need a `recuse_if` method because you need to test again whether `from_data` is a `Hash`.
Consider
```
recurse_data = from_data.kind_of?(Hash) ? from_data.values : from_data
recurse_data.each { |value| get_file_contents(value, base_url) } if recurse_data.kind_of?(Array)
```

Reply:
I think it is more flexible/readable having a single function to decide, but your implementation is so clever :D

I changed it and added a small comment to make it more straightforward.